Allow an AppMsgStake tx to change the address of a staked app

msmania opened the pull request:
The patch implements [PIP-35](https://forum.pokt.network/t/pip-35-introduce-a-secure-way-to-transfer-a-staked-app-to-a-new-account/4806), allowing the existing AppMsgStake transaction to change the address of a staked app.  This enables us to *transfer* the existing app slot from one to a new account without unstaking.  To make this operation easier, this patch also introduces a new command `app transfer`.

## Description

<!-- reviewpad:summarize:start -->
### Summary generated by Reviewpad on 15 Dec 23 11:53 UTC
This pull request includes changes in multiple files. Here is a summary of the changes:

1. The file `appStateChanges_test.go` has changes related to application state changes and transfers. The changes include the addition of new import statements, test functions, and helper functions.

2. The file `types.go` has changes related to the `MsgStake` type. The changes involve adding a new method `IsValidTransfer` to check for a special case in the `MsgStake` type where ownership transfer is being done.

3. The file `codec.go` has changes related to the support of an "App Transfer" feature. The changes include adding a constant and a new function for checking if a certain upgrade has occurred.

4. The file `expectedKeepers.go` has changes related to the addition of a new interface `AppKeeper`.

5. The file `keeper.go` has changes that involve adding a new field to the `Keeper` struct.

6. The file `baseapp.go` has changes that include formatting changes in the package comment and modifications in the `DeliverTx` function.

7. The file `app.go` has changes related to the assignment of a field in the `app.accountKeeper` object.

8. The file `auth.go` has changes that include adding comments and conditions for non-custodial and output address editor upgrades.

9. The file `common_test.go` has changes that involve importing packages, renaming a package, and adding/modifying functions.

10. The file `appStateChanges.go` has changes related to the validation and transfer functionality of applications.

11. The file `txUtil.go` has changes that add a new function for transferring an application.

12. The file `app/cmd/cli/app.go` has changes related to the addition of a new command for transferring the ownership of a staked app.

13. The file `keeper.go` has changes that add a new method for checking if a message is for transferring ownership.

14. The file `handler.go` has changes that include import statements, function parameter modifications, and logic for transferring application ownership.

These are the summaries of the changes in each file. Let me know if you have any specific questions or need further information regarding these changes.
<!-- reviewpad:summarize:end -->

POKT-Discourse commented:
This pull request has been mentioned on **Pocket Network Forum**. There might be relevant details there:

https://forum.pokt.network/t/pip-35-introduce-a-secure-way-to-transfer-a-staked-app-to-a-new-account/4806/1

msmania committed:
Introduce AppTransfer feature

The patch allows the existing AppMsgStake transaction to change the address of
a staked app, which enables us to *transfer* the existing app slot from one to
a new account without unstaking.

The AppMsgStake message must be signed by the current staked app.  To issue
such a transaction easier, this patch introduces a new `apps transfer` command
that takes the current app's address and a new app's public key.

POKT-Discourse commented:
This pull request has been mentioned on **Pocket Network Forum**. There might be relevant details there:

https://forum.pokt.network/t/era-cycle-report-cycle-2/4828/1

Olshansk commented on app/cmd/cli/app.go:
```diff
@@ -116,6 +117,65 @@ Prompts the user for the <fromAddr> account passphrase.`,
+	Use:   "transfer <fromAddr> <newAppPubKey> <networkID> <fee> [memo]",
+	Short: "Transfer the ownership of a staked app from one to another",
+	Long: `Submits a transaction to transfer the ownership of a staked app from
+<fromAddr> to a neww account specified as <newAppPubKey> without unstaking
```
```suggestion
<fromAddr> to a new account specified as <newAppPubKey> without unstaking
```

Olshansk commented on baseapp/baseapp.go:
```diff
@@ -810,9 +809,8 @@ func (app *BaseApp) DeliverTx(req abci.RequestDeliverTx) (res abci.ResponseDeliv
-		if signerPK == nil || messageType == types2.MsgAppStakeName {
-			signers := msg.GetSigners()
-			if len(signers) >= 1 {
+		if signerPK == nil {
```
Do you know why we used to have `messageType == types2.MsgAppStakeName`?

msmania replied:
Unfortunately no.  This was introduced by the large, combined commit  32157db43c88fcd37d5c68978f0256c8631db833 of v0.8, but this part does nothing meaningful.  MsgAppStake's `GetSigners` always returns a single element array and its value is `msg.PubKey` which was guaranteed to be the signer, so it doesn't make any sense to have this.

Olshansk replied:
Logically, I agree that you're right and I cannot prove otherwise. However, I'm really paranoid that there's some weird edge case and I don't want any of us to deal with a chain halt.

`Hypothesis`: When the message type is NOT an AppStake (e.g. delegated staking for suppliers / node providers), maybe we need to enforce to using the first signer?

msmania replied:
This patch does not change the logic in the case where the message type is not an AppStake.  Regardless of this patch, we enforce to using the first signer only if `signerPK` is null.  Moreover, this `signer` affects `abci.ResponseDeliverTx` which just goes to the transaction log.  I don't think it causes a chain halt.

Olshansk commented on codec/codec.go:
```diff
@@ -57,6 +57,7 @@ const (
 	VEDITKey                     = "VEDIT"
 	OutputAddressEditKey         = "OEDIT"
 	ClearUnjailedValSessionKey   = "CRVAL"
+	AppTransferKey               = "AppTransfer"
```
NIT: s/AppTransfer/APPTR?

msmania replied:
This may be bikeshedding.  Do we need to keep feature names cryptic codenames like those?  In my understanding, there is no limitation in case sensitivity or the length.  I think having more descriptive names makes the result of `/v1/query/upgrade` easier to understand.

Olshansk replied:
Not adamant either way. Feel free to resolve.

Olshansk commented on x/apps/handler.go:
```diff
@@ -30,25 +31,38 @@ func NewHandler(k keeper.Keeper) sdk.Handler {
+			return err.Result()
+		}
+	} else {
+		// then check if they can stake
```
```suggestion
		// otherwise check if the msg is to stake an application
```

Olshansk commented on x/apps/handler.go:
```diff
@@ -30,25 +31,38 @@ func NewHandler(k keeper.Keeper) sdk.Handler {
+	} else {
+		// then check if they can stake
+		if err := k.ValidateApplicationStaking(ctx, application, msg.Value); err != nil {
+			ctx.Logger().Error(fmt.Sprintf("Validate App Can Stake Error, at height: %d with address: %s", ctx.BlockHeight(), sdk.Address(addr).String()))
```
```suggestion
			ctx.Logger().Error(fmt.Sprintf("ERROR: ValidateApplicationStaking at height: %d with address: %s", ctx.BlockHeight(), sdk.Address(addr).String()))
```

msmania replied:
This error log is not new.  Let's leave it as is, though it has strange capitalization.

Olshansk commented on x/apps/handler.go:
```diff
@@ -30,25 +31,38 @@ func NewHandler(k keeper.Keeper) sdk.Handler {
+		}
 	}
 	// create the event
+	senderAddrStr := sdk.Address(signer.Address()).String()
```
s/senderAddrStr/signedAddrStr

msmania replied:
changed it to `signerAddrStr`

Olshansk commented on x/apps/handler.go:
```diff
@@ -30,25 +31,38 @@ func NewHandler(k keeper.Keeper) sdk.Handler {
-		return err.Result()
+	// check if the msg is to transfer an app first
+	if curApp, err := k.ValidateApplicationTransfer(ctx, signer, msg); err == nil {
+		if err := k.TransferApplication(ctx, curApp, msg.PubKey); err != nil {
```
Can you add a log line about here similar to the one you have below for consistency?

msmania replied:
I realized `TransferApplication` never fails.  Removed the return value and moved info logging from `TransferApplication` to this function.

Olshansk commented on x/apps/types/msg.go:
```diff
@@ -62,6 +66,24 @@ func (msg MsgStake) ValidateBasic() sdk.Error {
+	}
+
+	// The chains must be empty for transfer
+	if !msg.Value.IsZero() {
```
This is the same check as on line 75

msmania replied:
Good catch. Changed it to `len(msg.Chains) > 0`.

Olshansk replied:
Can you add a unit test to catch this please?

Olshansk commented on x/apps/keeper/appUtil.go:
```diff
@@ -30,3 +30,33 @@ func (k Keeper) AllApplications(ctx sdk.Ctx) (apps []exported.ApplicationI) {
 	return apps
 }
+
+func (k Keeper) IsMsgAppTransfer(
```
Please add a godoc comment

Olshansk commented on x/apps/types/msg.go:
```diff
@@ -45,6 +45,10 @@ func (msg MsgStake) GetSignBytes() []byte {
 
 // ValidateBasic quick validity check for staking an application
 func (msg MsgStake) ValidateBasic() sdk.Error {
+	if err := msg.IsValidTransfer(); err == nil {
```
Can you add a comment explaining that we're reusing the `MsgStake` command for transfers

Olshansk commented on x/apps/keeper/appStateChanges.go:
```diff
@@ -77,6 +101,40 @@ func (k Keeper) ValidateEditStake(ctx sdk.Ctx, currentApp types.Application, amo
+	if !ctx.IsAfterUpgradeHeight() ||
+		!k.Cdc.IsAfterAppTransferUpgrade(ctx.BlockHeight()) {
+		err = types.ErrApplicationStatus(k.codespace)
+		return
```
NIT: Can we use explicit return types (not named). It's harder to follow this way.

Olshansk commented on x/apps/keeper/appStateChanges.go:
```diff
@@ -77,6 +101,40 @@ func (k Keeper) ValidateEditStake(ctx sdk.Ctx, currentApp types.Application, amo
+	}
+
+	err = ensurePubKeyTypeSupported(ctx, msg.PubKey, k.Codespace())
+	if err != nil {
```
Single line

Olshansk commented on x/apps/keeper/appStateChanges.go:
```diff
@@ -77,6 +101,40 @@ func (k Keeper) ValidateEditStake(ctx sdk.Ctx, currentApp types.Application, amo
+		return
+	}
+
+	// The pubkey in the message must not exist
```
We explicitly set the `PubKey` in `MsgStake` when doing a transfer like so:

```go
	msg := appsType.MsgStake{PubKey: newAppPubKey}
	if err = msg.ValidateBasic(); err != nil {
		return nil, err
	}
```

A bit confused by this comment.

Do you mean, the `The pubKey of msgStake must not be a pre-existing application`?

msmania replied:
Yes, you're right.  What I meant was "not exist in the world state".  Applying your suggestion.

Olshansk commented on x/auth/ante.go:
```diff
@@ -72,6 +72,17 @@ func ValidateTransaction(ctx sdk.Ctx, k Keeper, stdTx types.StdTx, params Params
+		// a message where the pubkey is different from the signer, delegating
+		// most of validation work to the app's message handler after fee deduction.
+		if k.AppKeeper.IsMsgAppTransfer(ctx, msgSigner, stdTx.Msg) {
+			validSigners = append(validSigners, msgSigner)
```
Is `msgSigner` not part of `validSigners` by default?

msmania replied:
Counterintuitively it's not.  `stdTx.GetSigners()` is simply redirected to `tx.GetMsg().GetSigners()`, not looking at the actual signer.  This is why every transaction message has a redundant field storing its signer.

Olshansk replied:
Can you #PUC (#PleaseUpdateComment) with this additional detail & context?

Olshansk commented on x/apps/keeper/appStateChanges_test.go:
```diff
@@ -371,3 +374,137 @@ func TestAppStateChange_BeginUnstakingApplication(t *testing.T) {
+
+	ctx, _, keeper := createTestInput(t, true)
+
+	apps := make([]types.Application, 3)
```
```suggestion
	// Create and stake 3 apps
	apps := make([]types.Application, 3)
```

Olshansk commented on x/apps/keeper/appStateChanges_test.go:
```diff
@@ -371,3 +374,137 @@ func TestAppStateChange_BeginUnstakingApplication(t *testing.T) {
+
+	// apps[0]: staked
+	// apps[1]: unstaking
+	// apps[2]: staked and jailed
```
```suggestion
	// apps[2]: staked and jailed
	// apps[3]: not an application (see pubKeys[3]) and will be used for the transfer
```

Olshansk commented on x/apps/keeper/appStateChanges_test.go:
```diff
@@ -371,3 +374,137 @@ func TestAppStateChange_BeginUnstakingApplication(t *testing.T) {
+	assert.Equal(t, keeper.Codespace(), err.Codespace())
+	assert.Equal(t, types.CodeInvalidStatus, err.Code())
+
+	// transfer: apps[3]-->apps[0](previous owner) success
```
1. Why do we expect this to succeed?
2. If we were to wait for a new block, would this fail since the previous app is never unstaked?

msmania replied:
`transferApp` directly modifies the world state.  There is no block generation.  This means at line 468, the app has been transferred to `apps[3]` and `apps[0]` is not a staked app.  That's why we can transfer the ownership back to `apps[0]`.

Olshansk replied:
Can you update the test to verify that IF we do wait for a block, the transfer back fails?

msmania replied:
> Can you update the test to verify that IF we do wait for a block, the transfer back fails?

If a block was generated after the line 462, the test would still pass.  After the app ownership is transferred from A to B, the address A is the same as being unstaked or never being staked, meaning the address can be staked again.

Olshansk commented on x/apps/keeper/appStateChanges_test.go:
```diff
@@ -371,3 +374,137 @@ func TestAppStateChange_BeginUnstakingApplication(t *testing.T) {
+	assert.True(t, stakedApp.IsJailed())
+
+	// transfer a jailed app - success
+	err = transferApp(t, &ctx, &keeper, pubKeys[2], pubKeys[3])
```
pubKeys[3] was already transfered to above. If we were to wait for a new block, would this fail?

msmania replied:
Same here.  At line 502, the app ownership has been transferred back to `apps[0]`.  `apps[3]` is not a staked app.

Olshansk reviewed (changes requested):
Nothing blocking, but just a couple nits and questions.

msmania committed:
address comments

Olshansk reviewed (changes requested):
Added a few followup comments to the previous review. PTAL

Olshansk commented:
@msmania See the one linger comment I have [here](https://github.com/pokt-network/pocket-core/pull/1585#discussion_r1428569418).

--- 

Also, I tried to do a `Show changes since your last review` (see screenshots below) but I'm guessing there was a rebase/force-push/etc done to the branch which prevents that from happening.

I'm travelling on a trip this weekend and will aim to take a second look at this mid next week.

<img width="502" alt="Screenshot 2023-12-15 at 3 18 11 PM" src="https://github.com/pokt-network/pocket-core/assets/1892194/b9e8fecf-4ce9-492d-b5f2-b2dbcabac70c">

<img width="1192" alt="Screenshot 2023-12-15 at 3 19 06 PM" src="https://github.com/pokt-network/pocket-core/assets/1892194/d82f586f-fe35-478e-af77-f25707044baa">

Olshansk reviewed (approved):
Thanks @msmania!

POKT-Discourse commented:
This pull request has been mentioned on **Pocket Network Forum**. There might be relevant details there:

https://forum.pokt.network/t/rc-0-11-1-upgrade-and-hi/5012/1
